pkg/registry/cache.go

```diff
@@ -0,0 +1,497 @@
+/*
+Copyright The Helm Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package registry // import "k8s.io/helm/pkg/registry"
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"sort"
+	"strings"
+	"time"
+
+	orascontent "github.com/deislabs/oras/pkg/content"
+	"github.com/docker/go-units"
+	checksum "github.com/opencontainers/go-digest"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	"github.com/pkg/errors"
+
+	"k8s.io/helm/pkg/chart"
+	"k8s.io/helm/pkg/chart/loader"
+	"k8s.io/helm/pkg/chartutil"
+)
+
+type (
+	filesystemCache struct {
+		out     io.Writer
+		rootDir string
+		store   *orascontent.Memorystore
+	}
+)
+
+func (cache *filesystemCache) LayersToChart(layers []ocispec.Descriptor) (*chart.Chart, error) {
+	metaLayer, contentLayer, err := extractLayers(layers)
+	if err != nil {
+		return nil, err
+	}
+
+	name, version, err := extractChartNameVersionFromLayer(contentLayer)
+	if err != nil {
+		return nil, err
+	}
+
+	// Obtain raw chart meta content (json)
+	_, metaJSONRaw, ok := cache.store.Get(metaLayer)
+	if !ok {
+		return nil, errors.New("error retrieving meta layer")
+	}
+
+	// Construct chart metadata object
+	metadata := chart.Metadata{}
+	err = json.Unmarshal(metaJSONRaw, &metadata)
+	if err != nil {
+		return nil, err
+	}
+	metadata.Name = name
+	metadata.Version = version
+
+	// Obtain raw chart content
+	_, contentRaw, ok := cache.store.Get(contentLayer)
+	if !ok {
+		return nil, errors.New("error retrieving meta layer")
+	}
+
+	// Construct chart object and attach metadata
+	ch, err := loader.LoadArchive(bytes.NewBuffer(contentRaw))
+	if err != nil {
+		return nil, err
+	}
+	ch.Metadata = &metadata
+
+	return ch, nil
+}
+
+func (cache *filesystemCache) ChartToLayers(ch *chart.Chart) ([]ocispec.Descriptor, error) {
+
+	// extract/separate the name and version from other metadata
+	if ch.Metadata == nil {
+		return nil, errors.New("chart does not contain metadata")
+	}
+	name := ch.Metadata.Name
+	version := ch.Metadata.Version
+
+	// Create meta layer, clear name and version from Chart.yaml and convert to json
+	ch.Metadata.Name = ""
+	ch.Metadata.Version = ""
+	metaJSONRaw, err := json.Marshal(ch.Metadata)
+	if err != nil {
+		return nil, err
+	}
+	metaLayer := cache.store.Add(HelmChartMetaFileName, HelmChartMetaMediaType, metaJSONRaw)
+
+	// Create content layer
+	// TODO: something better than this hack. Currently needed for chartutil.Save()
```

Comment:
what is happening here?  can you add a comment about why this is required

Reply:
Added some more comments in new commit, basically the Name and Version are required to save a chart. We need to rethink the relationship between metadata and chart content